### PR TITLE
[RELAY][OP] Split

### PR DESCRIPTION
**Pull request by srkreddy1238**
Thanks for contributing to TVM!   Please refer to guideline https://docs.tvm.ai/contribute/ for useful information and tips. After the pull request is submitted, please request code reviews from others in the community.

### CODE REVIEW

File: src/relay/op/tensor/transform.cc [hunk 1]
```diff
@@ -404,5 +404,92 @@ Examples::
...
+
+bool SplitRel(const Array<Type>& types,
+              int num_inputs,
+              const Attrs& attrs,
+              const TypeReporter& reporter) {
+  // `types` contains: [data, result]
+  CHECK_EQ(types.size(), 2);
+  const auto* data = types[0].as<TensorTypeNode>();
+  CHECK(data != nullptr);
+  CHECK_NE(data->shape.size(), 0);
```
Comment (junrushao1994) [inline src/relay/op/tensor/transform.cc]: Could we add more detailed error message?

File: src/relay/op/tensor/transform.cc [hunk 2]
```diff
@@ -404,5 +404,92 @@ Examples::
...
+  const auto param = attrs.as<SplitAttrs>();
+  CHECK(param != nullptr);
+
+  if (param->equal_split) {
+    const auto num_outputs = as_const_int(param->indices_or_sections[0]);
+    CHECK_LT(param->axis, data->shape.size());
```
Comment (junrushao1994) [inline src/relay/op/tensor/transform.cc]: Same above

File: include/tvm/relay/attrs/transform.h [hunk 3]
```diff
@@ -68,6 +68,21 @@ struct TakeAttrs : public tvm::AttrsNode<TakeAttrs> {
...
 
+struct SplitAttrs : public tvm::AttrsNode<SplitAttrs> {
+  Array<IndexExpr> indices_or_sections;
+  int axis;
+  bool equal_split;
+
+  TVM_DECLARE_ATTRS(SplitAttrs, "relay.attrs.SplitAttrs") {
+    TVM_ATTR_FIELD(indices_or_sections)
+        .describe("Number of outputs to be splitted");
+    TVM_ATTR_FIELD(axis).set_lower_bound(0).set_default(1)
```
Comment (junrushao1994) [inline include/tvm/relay/attrs/transform.h]: nit: numpy has axis defaults to 0, and does not seem to enforce axis to be nonnegative.

https://docs.scipy.org/doc/numpy/reference/generated/numpy.split.html

**Comment by srkreddy1238** [comment]
cc @junrushao1994 please have another look.

File: src/relay/op/tensor/transform.cc [hunk 4]
```diff
@@ -498,5 +498,97 @@ and type as the input array.
...
+    << "axis should be within the input dimension range.";
+  CHECK_GT(axis, 0)
+    << "axis should be within the input dimension range.";
+
+  if (param->equal_split) {
+    const auto num_outputs = as_const_int(param->indices_or_sections[0]);
+    // TODO(Siva) Reuse Assert from other PR.
+    // CHECK(reporter->Assert(data->shape[axis] %
+    //                       param->indices_or_sections[0] == make_zero(Int(64))))
+    //      << "indices_or_sections need to be able to divide input.shape[axis]";
```
Comment (yzhliu) [inline src/relay/op/tensor/transform.cc]: if it is required, we'd better uncomment and change it later.

File: python/tvm/relay/op/transform.py [hunk 5]
```diff
@@ -180,3 +180,28 @@ def full_like(data, fill_value):
...
+        The source array.
+
+    indices_or_sections : tuple of int
+        The indices or number of sections to split.
+
+    axis : int, optional
+        The axis over which to split.
+
+    equal_split : boolean
+        To split into indices_or_section[0] sections.
```
Comment (yzhliu) [inline python/tvm/relay/op/transform.py]: could you explain a bit more for `indices_or_sections` and `equal_split`? it seems to be not so clear to users.

File: python/tvm/relay/op/transform.py [hunk 6]
```diff
@@ -242,3 +242,34 @@ def where(condition, x, y):
...
+
+    Returns
+    -------
+    ret : relay.Expr
+        The computed result.
+    """
+    return _make.split(data, indices_or_sections, axis, equal_split)
```
Comment (tqchen) [inline python/tvm/relay/op/transform.py]: make use of https://github.com/dmlc/tvm/blob/master/python/tvm/relay/expr.py#L216

**Comment by tqchen** [comment]
@junrushao1994 @srkreddy1238 can you followup again on this?

File: src/relay/op/tensor/transform.cc [hunk 7]
```diff
@@ -701,5 +701,100 @@ RELAY_REGISTER_OP("squeeze")
...
+      // CHECK(reporter->Assert(param->indices_or_sections[i] > begin))
+      //    << "indices_or_sections need to be a sorted ascending list";
+      std::vector<IndexExpr>&& oshape = AsVector(data->shape);
+      oshape[axis] = param->indices_or_sections[i] - begin;
+      begin = param->indices_or_sections[i];
+      auto vec_type = TensorTypeNode::make(oshape, data->dtype);
+      fields.push_back(vec_type);
+    }
+    // CHECK(reporter->Assert(begin < data->shape[axis]))
+    //    << "The sum of sections must match the input.shape[axis]";
```
Comment (junrushao1994) [inline src/relay/op/tensor/transform.cc]: Should we uncomment or delete these commented codes?

**Comment by tqchen** [comment]
please rebase against master and note the changes in #1934

**Comment by srkreddy1238** [comment]
@tqchen have a look

File: python/tvm/relay/op/transform.py [hunk 8]
```diff
@@ -242,3 +243,35 @@ def where(condition, x, y):
     Note that the shape of condition, x, and y needs to be the same.
     """
     return _make.where(condition, x, y)
+
+
+def split(data, indices_or_sections, axis=0, equal_split=False):
```
Comment (tqchen) [inline python/tvm/relay/op/transform.py]: as per numpy convention, indices_or_sections
 
- can either be an integer(which indicates equal split)
-  or Array of Int, which(indicate the sections)

We want the API behavior to be as consistent as possible. In order to do so, we can store the indices_or_sections as NodeRef in the Attrs.  Now, in the _make.split C++ API, we should check if the input type is kDLInt, and convert that to IntImm.

During type inference, we can check if the indices_or_sections is IntImm, and decide accordingly.

Comment (tqchen) [reply python/tvm/relay/op/transform.py]: This does make the code a bit harder to read and we should explain clearly by commenting on the field of indices_or_sections in the Attrs, to say the two expected field, and the reason of doing so(compatible with numpy convention). We will later try to introduce a Union type which might make it more readable.

File: src/relay/op/tensor/transform.cc [hunk 9]
```diff
@@ -779,5 +781,103 @@ RELAY_REGISTER_OP("broadcast_to_like")
...
+  if (axis < 0) {
+    axis += data->shape.size();
+  }
+  CHECK_LT(axis, data->shape.size())
+    << "axis should be within the input dimension range.";
+  CHECK_GT(axis, 0)
+    << "axis should be within the input dimension range.";
+
+  if (param->indices_or_sections.as<IntImm>()) {
+    const auto sections = make_const(Int(32),
```
Comment (tqchen) [inline src/relay/op/tensor/transform.cc]: you can directly use param->indices_or_sections without making another const?

File: src/relay/op/tensor/transform.cc [hunk 10]
```diff
@@ -779,5 +781,103 @@ RELAY_REGISTER_OP("broadcast_to_like")
...
+  auto axis = param->axis;
+  if (axis < 0) {
+    axis += data->shape.size();
+  }
+  CHECK_LT(axis, data->shape.size())
+    << "axis should be within the input dimension range.";
+  CHECK_GT(axis, 0)
+    << "axis should be within the input dimension range.";
+
+  if (param->indices_or_sections.as<IntImm>()) {
```
Comment (tqchen) [inline src/relay/op/tensor/transform.cc]: you can do const IntImm* op =  here so you can get the value later(without using another as_const_int

File: src/relay/op/tensor/transform.cc [hunk 11]
```diff
@@ -779,5 +781,103 @@ RELAY_REGISTER_OP("broadcast_to_like")
...
+        auto vec_type = TensorTypeNode::make(oshape, data->dtype);
+        fields.push_back(vec_type);
+    }
+    reporter->Assign(types[1], TupleTypeNode::make(Array<Type>(fields)));
+  } else {
+    auto indices = param->indices_or_sections.as<ArrayNode>()->data;
+    const auto num_outputs = indices.size() + 1;
+    auto begin = IndexExpr(make_zero(Int(32)));
+    std::vector<Type> fields;
+    for (uint i = 0; i < num_outputs - 1; ++i) {
```
Comment (tqchen) [inline src/relay/op/tensor/transform.cc]: indices.size()

File: tests/python/relay/test_op_level3.py [hunk 12]
```diff
@@ -99,6 +99,37 @@ def verify_take(dshape, indices_shape, oshape, axis=None):
     verify_take((d1, d2), (d3, d4, d5), (d1, d3, d4, d5), 1)
     verify_take((d1, d2, d3, d4), (d5, d6), (d1, d2, d5, d6, d4), -2)
 
+def test_split_infer_type():
+    def verify_split(dshape, indices_or_sections, ret_type, axis=None):
+        x = relay.var("x", relay.ty.TensorType(dshape, "float32"))
+        y = relay.split(x, indices_or_sections, axis=axis)
```
Comment (tqchen) [inline tests/python/relay/test_op_level3.py]: please add a astext() testcase to check if text printer behaves correctly.

**Comment by srkreddy1238** [comment]
cc @tqchen pls have a look at the alpha equal for ```Div``` operator too.

**Comment by srkreddy1238** [comment]
cc @tqchen

**Comment by tqchen** [comment]
Thanks @srkreddy1238 @junrushao1994 @yzhliu this is merged

**Pull request merged**